src/django_components/util/context.py

```diff
@@ -0,0 +1,114 @@
+import copy
+from typing import List
+
+from django.template import Context
+from django.template.loader_tags import BlockContext
+
+
+class CopiedDict(dict):
+    """Dict subclass to identify dictionaries that have been copied with `snapshot_context`"""
+
+    pass
+
+
+def snapshot_context(context: Context) -> Context:
```

Comment:
This is the implementation of logic to make a Context "snapshot".

Basically it copies the Context object, and all the dictionaries in it. But to avoid too much copying, the dictionaries are copied only once. And to detect, if a dict was already copied or not, they are instantiated as `CopiedDict`, which is a subclass of `dict`, and so behaves the same way.

I copy not the list at `Context.dicts`, but also individual dictionaries out of precaution, because I'm not 100% sure e.g. how nested forloops behave - whether they make copies of the parent loop's data, or mutate the existing object. And in case of the latter, we want to copy the dicts.

Reply:
I tested it just now, and we do indeed need to copy the dicts.